sru/modules.py

```diff
@@ -0,0 +1,654 @@
+
+    def forward(self, input: Tensor,
+                      c0: Optional[Tensor] = None,
+                      mask_pad: Optional[Tensor] = None) -> Tuple[Tensor, Tensor]:
+        """The forward method of SRU module
+
+        Parameters
+        ----------
+        input: Tensor
+            the input feature. shape: (length, batch_size, input_size)
+        c0: Tensor, optional
+            the initial internal hidden state. shape: (num_layers,
+            batch_size, output_size) where
+            output_size = hidden_size * num_direction
+        mask_pad: Tensor, optional
+            the mask where a non-zero value indicates if an input token
+            is pad token that should be ignored in forward and backward
+            computation. shape: (length, batch_size)
+
+        Returns
+        ----------
+        h: Tensor
+            the output hidden state. shape: (length, batch_size,
+            output_size) where
+            output_size = hidden_size * num_direction
+        c: Tensor
+            the last internal hidden state. shape: (num_layers,
+            batch_size, output_size), or (num_layers * num_directions,
+            batch_size, hidden_size) if `nn_rnn_compatible_return` is
+            set `True`
+
+        """
+        # unpack packed, if input is packed. packing and then unpacking will be slower than not packing
+        # at all, but makes SRU usage compatible with nn.RNN usage
+        orig_input = input
+        if isinstance(orig_input, PackedSequence):
+            input, batch_sizes, sorted_indices, unsorted_indices = input
+            length = input.size(0)
+            batch_size = input.size(1)
+            mask_pad = torch.arange(batch_size, device=batch_sizes.device).expand(length, batch_size)
+            mask_pad = (mask_pad >= batch_sizes.view(length, 1)).contiguous()
+        else:
+            length = input.size(0)
+            batch_size = input.size(1)
+            batch_sizes = None
+            sorted_indices = None
+            unsorted_indices = None
+
+        # The dimensions of `input` should be: `(sequence_length, batch_size, input_size)`.
+        if input.dim() != 3:
+            raise ValueError("There must be 3 dimensions for (length, batch_size, input_size)")
+
+        if c0 is None:
+            zeros = torch.zeros(input.size(1), self.output_size, dtype=input.dtype,
+                                device=input.device)
+            c0_ = [ zeros for i in range(self.num_layers) ]
+        else:
+            # The dimensions of `c0` should be: `(num_layers, batch_size, hidden_size * dir_)`.
+            if c0.dim() != 3:
+                raise ValueError("There must be 3 dimensions for (num_layers, batch_size, output_size)")
+            c0_ = [ x.squeeze(0) for x in c0.chunk(self.num_layers, 0) ]
+
+        if self.input_to_hidden is None:
+            prevx = input
+        else:
+            prevx = self.input_to_hidden(input)
+        lstc = []
+        i = 0
```

Comment:
I see in one of the commits you reverted using `enumerate` here? I think I prefer the `enumerate` version on the whole, since shorter, easier to read. Is there any reason why the `enumerate` version was reverted?